File: frontend/src/types/forge.ts
// src/types/forge.ts
// PR-window DTO shapes, mirroring src/api/schemas/forge.py.
// Hand-written pending openapi-ts coverage of the control-plane routes.

export interface Person {
  name: string;
  email: string | null;
  login: string | null;
  avatar_url: string | null;
}

export interface CommitNode {
  sha: string;
  parents: string[];
  summary: string;
  author: Person;
  committer: Person;
  authored_at: string;
  committed_at: string;
  refs: string[];
  pr_number: number | null;
}

export interface CommitGraph {
  nodes: CommitNode[];
  source: string;
  truncated: boolean;
  head_sha: string | null;
  dangling_parents: string[];
}

export type PrStateValue = 'open' | 'draft' | 'merged' | 'closed';

export interface PullRequest {
  number: number;
  title: string;
  state: PrStateValue;
  head_ref: string;
  base_ref: string;
  head_sha: string;
  author: Person;
  review_state: string;
  checks: string;
  requested_reviewers: string[];
  is_mergeable: boolean | null;
  created_at: string;
  updated_at: string;
  source: string;
}

export interface ForgeCapabilities {
  source: string;
  supports_prs: boolean;
  supports_reviews: boolean;
  supports_checks: boolean;
}

File: src/domain/value_objects/forge.py
"""
src/domain/value_objects/forge.py — git-forge projection value objects.

Read-only projections of git/forge state for the PR window. Three data tiers:
  1. commit topology (both GitHub and local git can populate) — CommitNode;
  2. PRs / reviews / checks (GitHub only) — PullRequest;
  3. identity that differs by source (GitHub login vs git name+email) — Person.

These are projections: immutable, never written to the task tables. A ``source``
discriminator + ``ForgeCapabilities`` let a local-git adapter advertise the gaps
(no PRs) instead of throwing.
"""
from __future__ import annotations

from datetime import datetime
from enum import Enum

from pydantic import BaseModel, ConfigDict


class _Frozen(BaseModel):
    model_config = ConfigDict(frozen=True)


class DataSource(str, Enum):
    GITHUB = "github"
    LOCAL_GIT = "local_git"


class Person(_Frozen):
    name: str
    email: str | None = None
    login: str | None = None        # GitHub only
    avatar_url: str | None = None   # GitHub only


class CommitNode(_Frozen):
    sha: str                        # full 40-char always
    parents: tuple[str, ...]        # 0 = root, 1 = normal, 2+ = merge
    summary: str
    author: Person
    committer: Person
    authored_at: datetime
    committed_at: datetime
    refs: tuple[str, ...] = ()
    pr_number: int | None = None    # None from local git


class CommitGraph(_Frozen):
    nodes: tuple[CommitNode, ...]
    source: DataSource
    truncated: bool
    head_sha: str | None
    dangling_parents: frozenset[str] = frozenset()


class ReviewState(str, Enum):
    APPROVED = "approved"
    CHANGES_REQUESTED = "changes_requested"
    COMMENTED = "commented"
    PENDING = "pending"
    DISMISSED = "dismissed"


class CheckState(str, Enum):
    SUCCESS = "success"
    FAILURE = "failure"
    PENDING = "pending"
    NEUTRAL = "neutral"
    UNKNOWN = "unknown"


class PrState(str, Enum):
    OPEN = "open"
    DRAFT = "draft"
    MERGED = "merged"
    CLOSED = "closed"


class PullRequest(_Frozen):
    number: int
    title: str
    state: PrState
    head_ref: str
    base_ref: str
    head_sha: str
    author: Person
    review_state: ReviewState
    checks: CheckState
    requested_reviewers: tuple[str, ...]
    is_mergeable: bool | None        # None = not yet known ("checking…")
    created_at: datetime
    updated_at: datetime
    source: DataSource


class ForgeCapabilities(_Frozen):
    source: DataSource
    supports_prs: bool
    supports_reviews: bool
    supports_checks: bool
